src/esw/science_board/science_board.py

```diff
@@ -0,0 +1,587 @@
+
+The science_board codebase deals with reading and parsing NMEA like messages
+from the STM32 chip on the science PCB over UART to complete tasks for almost
+every mission. These tasks include operating the science box and getting
+relevant data during the Science task, controlling the arm laser during the
+Equipment Servicing task, and controlling the LED array used during the
+Autonomous Traversal task.
+'''
+import threading
+from typing import Any, Callable
+
+import numpy as np
+import rospy
+import serial
+from mrover.msg import Enable, Heater, Spectral, Thermistor
+from mrover.srv import (ChangeAutonLEDState, ChangeAutonLEDStateRequest,
+                        ChangeAutonLEDStateResponse, ChangeDeviceState,
+                        ChangeDeviceStateRequest, ChangeDeviceStateResponse,
+                        ChangeHeaterState, ChangeHeaterStateRequest,
+                        ChangeHeaterStateResponse, ChangeServoAngles,
+                        ChangeServoAnglesRequest, ChangeServoAnglesResponse)
+
+
+class ScienceBridge():
+    """Manages all the information and functions used for dealing with the
+    bridge between the science board and the Jetson.
+
+    One ScienceBridge will be made in main.
+
+    Attributes:
+        _led_map: A dictionary that maps the possible colors to an integer for
+        the UART message.
+        _mosfet_dev_map: A dictionary that maps each actual device to a MOSFET
+        device number.
+        _nmea_handle_mapper: A dictionay that maps each NMEA tag for a UART
+        message to its corresponding callback function that returns a ROS
+        struct with the packaged data.
+        _nmea_publisher_mapper: A dictionary that maps each NMEA tag for a
+        UART message to its corresponding ROS Publisher object.
+        _sleep: A float representing the sleep duration used for when the
+        sleep function is called.
+        _uart_transmit_msg_len: An integer representing the maximum length for
+        a transmitted UART message.
+        _uart_lock: A lock used to prevent clashing over the UART transmit
+        line.
+    """
+    _led_map: dict[str, int]
+    _mosfet_dev_map: dict[str, int]
+    _nmea_handle_mapper: dict[str, Callable]
+    _nmea_publisher_mapper: dict[str, rospy.Publisher]
+    _sleep: float
+    _uart_transmit_msg_len: int
+    _uart_lock: threading.Lock
+
+    def __init__(self):
+        self._led_map = rospy.get_param("/science_board/led_map")
+        self._mosfet_dev_map = rospy.get_param(
+            "/science_board/mosfet_device_map")
+        self._nmea_handle_mapper = {
+            "AUTOSHUTOFF": self._heater_auto_shut_off_handler,
+            "HEATER": self._heater_state_handler,
+            "SPECTRAL": self._spectral_handler,
+            "THERMISTOR": self._thermistor_handler,
+            "TRIAD": self._triad_handler
+        }
+        self._nmea_publisher_mapper = {
+            "AUTOSHUTOFF": rospy.Publisher(
```

Comment:
for the publishers, can we do something like science/{topic}. Just so the topics are organized when you do rostopic list.

Reply:
Yea good idea